hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/rsql/RSQLUtility.java

```diff
@@ -659,34 +716,123 @@ private Predicate getEqualToPredicate(final Object transformedValue, final Path<
-                final char escapeChar) {
+        private Predicate toNullOrNotEqualPredicate(final Path<Object> fieldPath, final String sqlValue) {
+            return cb.or(cb.isNull(pathOfString(fieldPath)),
+                    cb.notLike(cb.upper(pathOfString(fieldPath)), sqlValue, ESCAPE_CHAR));
+        }
+
+        private Predicate toNotNullOrEmptyPredicate(final Path<Object> fieldPath) {
+            return cb.and(cb.isNotNull(pathOfString(fieldPath)), cb.notEqual(pathOfString(fieldPath), ""));
+        }
+
+        @SuppressWarnings({ "rawtypes", "unchecked" })
+        private Predicate toNotEqualWithSubQueryPredicate(final A enumField, final String sqlValue,
+                final String[] fieldNames) {
+
+            final Class<?> javaType = root.getJavaType();
+            final Subquery<?> subQuery = query.subquery(javaType);
+            final Root subQueryRoot = subQuery.from(javaType);
+
+            final Path<?> innerFieldPath = getInnerFieldPath(subQueryRoot, fieldNames, enumField.isMap());
+            final Expression<String> expressionToCompare = getExpressionToCompare(innerFieldPath, enumField);
+
+            subQuery.select(subQueryRoot)
+                    .where(cb.and(
+                            cb.equal(root.get(enumField.identifierFieldName()),
+                                    subQueryRoot.get(enumField.identifierFieldName())),
+                            cb.like(cb.upper(expressionToCompare), sqlValue)));
+
+            return cb.not(cb.exists(subQuery));
+        }
+
+        private static String[] getSubAttributesFrom(final String property) {
+            return property.split("\\" + FieldNameProvider.SUB_ATTRIBUTE_SEPERATOR);
+        }
+
+        private static boolean isSimpleField(final String[] split, final boolean isMapKeyField) {
+            return split.length == 1 || (split.length == 2 && isMapKeyField);
+        }
+
+        private Expression<String> getExpressionToCompare(final Path innerFieldPath, final A enumField) {
+            if (!enumField.isMap()) {
+                return pathOfString(innerFieldPath);
+            }
+            if (innerFieldPath instanceof MapJoin) {
+                // Currently we support only string key .So below cast
+                // is safe.
+                return (Expression<String>) (((MapJoin<?, ?, ?>) pathOfString(innerFieldPath)).value());
+            }
+            final String valueFieldName = enumField.getSubEntityMapTuple().map(Entry::getValue)
+                    .orElseThrow(() -> new UnsupportedOperationException(
+                            "For the fields, defined as Map, only Map java type or tuple in the form of SimpleImmutableEntry are allowed. Neither of those could be found!"));
+            return pathOfString(innerFieldPath).get(valueFieldName);
+        }
+
+        private static Path<?> getInnerFieldPath(final Root<?> subqueryRoot, final String[] split,
```

Comment:
seems similar to _getFieldPath_ method, could they be refactored to remove duplication/unify logic?

Reply:
Refactoring seems to be not possible as the methods are too different

Reply:
```suggestion
        private static Path<?> getInnerFieldPath(final Root<?> subqueryRoot, final String[] split,
                final boolean isMapKeyField) {
            return getFieldPath(subqueryRoot, split, isMapKeyField,
                    (fieldPath, fieldNameSplit) -> getInnerJoinFieldPath(subqueryRoot, fieldPath, fieldNameSplit));
        }

        private static Path<?> getInnerJoinFieldPath(final Root<?> subqueryRoot, final Path<?> fieldPath,
                final String fieldNameSplit) {
            if (fieldPath instanceof Join) {
                return subqueryRoot.join(fieldNameSplit, JoinType.INNER);
            }

            return fieldPath;
        }
```
Check if still working